clients/rust-legacy/tests/unwrap_lamports.rs

```diff
@@ -0,0 +1,338 @@
+            TransactionError::InstructionError(
+                0,
+                InstructionError::Custom(TokenError::OwnerMismatch as u32)
+            )
+        )))
+    );
+
+    // unwrap None lamports is ok
+    token
+        .unwrap_lamports(
+            &alice_account,
+            &bob_account,
+            &alice.pubkey(),
+            None,
+            &[&alice],
+        )
+        .await
+        .unwrap();
+}
+
+#[tokio::test]
+async fn basic() {
+    let mut context = TestContext::new().await;
+    context.init_token_with_native_mint().await.unwrap();
+    run_basic_unwrap_lamports(context, TestMode::Regular).await;
+}
+
+#[tokio::test]
+async fn basic_with_extensions() {
+    let mut context = TestContext::new().await;
+    context.init_token_with_native_mint().await.unwrap();
+    run_basic_unwrap_lamports(context, TestMode::WithImmutableOwner).await;
+}
+
+async fn run_self_unwrap_lamports(context: TestContext, test_mode: TestMode) {
+    let TokenContext { token, alice, .. } = context.token_context.unwrap();
+
+    let amount = 1000000000;
+
+    let alice_account = Keypair::new();
+    match test_mode {
+        TestMode::WithImmutableOwner => {
+            token
+                .wrap(
+                    &alice_account.pubkey(),
+                    &alice.pubkey(),
+                    amount,
+                    &[&alice_account],
+                )
+                .await
+                .unwrap();
+        }
+        TestMode::Regular => {
+            token
+                .wrap_with_mutable_ownership(
+                    &alice_account.pubkey(),
+                    &alice.pubkey(),
+                    amount,
+                    &[&alice_account],
+                )
+                .await
+                .unwrap();
+        }
+    }
+    let alice_account = alice_account.pubkey();
+
+    // unwrap Some(1) lamports is ok
+    token
+        .unwrap_lamports(
+            &alice_account,
+            &alice_account,
+            &alice.pubkey(),
+            Some(1),
+            &[&alice],
+        )
+        .await
+        .unwrap();
+
+    // unwrap too much lamports is not ok
+    let error = token
+        .unwrap_lamports(
+            &alice_account,
+            &alice_account,
+            &alice.pubkey(),
+            Some(amount),
+            &[&alice],
+        )
+        .await
+        .unwrap_err();
+    assert_eq!(
+        error,
+        TokenClientError::Client(Box::new(TransportError::TransactionError(
+            TransactionError::InstructionError(
+                0,
+                InstructionError::Custom(TokenError::InsufficientFunds as u32)
+            )
+        )))
+    );
+
+    // unwrap None lamports is ok
+    token
+        .unwrap_lamports(
+            &alice_account,
+            &alice_account,
+            &alice.pubkey(),
+            None,
+            &[&alice],
+        )
+        .await
+        .unwrap();
```

Comment:
Can we check that the account is properly a system account with the same amount of lamports?